Mathlib/Data/Finset/Sups.lean

```diff
@@ -0,0 +1,535 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+
+! This file was ported from Lean 3 source module data.finset.sups
+! leanprover-community/mathlib commit 008205aa645b3f194c1da47025c5f110c8406eab
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Finset.NAry
+import Mathlib.Tactic.ScopedNS
+
+/-!
+# Set family operations
+
+This file defines a few binary operations on `Finset α` for use in set family combinatorics.
+
+## Main declarations
+
+* `Finset.Sups s t`: Finset of elements of the form `a ⊔ b` where `a ∈ s`, `b ∈ t`.
+* `Finset.Infs s t`: Finset of elements of the form `a ⊓ b` where `a ∈ s`, `b ∈ t`.
+* `Finset.DisjSups s t`: Finset of elements of the form `a ⊔ b` where `a ∈ s`, `b ∈ t` and `a`
+  and `b` are disjoint.
+
+## Notation
+
+We define the following notation in locale `FinsetFamily`:
+* `s ⊻ t` for `Finset.Sups s t`
+* `s ⊼ t` for `Finset.Infs s t`
+* `s ○ t` for `Finset.DisjSups s t`
+
+## References
+
+[B. Bollobás, *Combinatorics*][bollobas1986]
+-/
+
+
+open Function
+
+variable {α : Type _} [DecidableEq α]
+
+namespace Finset
+
+section Sups
+
+variable [SemilatticeSup α] (s s₁ s₂ t t₁ t₂ u : Finset α)
+
+/-- The finset of elements of the form `a ⊔ b` where `a ∈ s`, `b ∈ t`. -/
+def Sups (s t : Finset α) : Finset α :=
+  image₂ (· ⊔ ·) s t
+#align finset.sups Finset.Sups
+
+-- mathport name: finset.sups
+scoped[FinsetFamily] infixl:74 " ⊻ " => Finset.Sups
+--porting note: had to open FinsetFamily here and import ScopedNS tactic
+open FinsetFamily
+
+variable {s t} {a b c : α}
+
+@[simp]
+theorem memSups : c ∈ s ⊻ t ↔ ∃ a ∈ s, ∃ b ∈ t, a ⊔ b = c := by simp [Sups]
+#align finset.mem_sups Finset.memSups
+
+variable (s t)
+
+@[simp, norm_cast]
+theorem coeSups : (s ⊻ t : Set α) = Set.image2 (· ⊔ ·) s t :=
+  coe_image₂ _ _ _
+#align finset.coe_sups Finset.coeSups
+
+theorem cardSups_le : (s ⊻ t).card ≤ s.card * t.card :=
+  card_image₂_le _ _ _
+#align finset.card_sups_le Finset.cardSups_le
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+theorem cardSups_iff :
+    (s ⊻ t).card = s.card * t.card ↔ (s ×ˢ t : Set (α × α)).InjOn fun x => x.1 ⊔ x.2 :=
+  card_image₂_iff
+#align finset.card_sups_iff Finset.cardSups_iff
+
+variable {s s₁ s₂ t t₁ t₂ u}
+
+theorem sup_memSups : a ∈ s → b ∈ t → a ⊔ b ∈ s ⊻ t :=
+  mem_image₂_of_mem
+#align finset.sup_mem_sups Finset.sup_memSups
+
+theorem sups_subset : s₁ ⊆ s₂ → t₁ ⊆ t₂ → s₁ ⊻ t₁ ⊆ s₂ ⊻ t₂ :=
```

Comment:
Should it be `@[mono]` (both here and in `mathlib3`)?

Reply:
Don't know, did @YaelDillies write that code?

Reply:
Possibly, yeah